File: core/src/main/scala/com/graphbrain/eco/nodes/FunNode.scala
package com.graphbrain.eco.nodes

abstract class FunNode(params: Array[ProgNode], lastTokenPos: Int= -1) extends ListNode(params, lastTokenPos) {
  val label: String

  override def toString = "(" + label + " " + params.mkString(" ") + ")"

  override def equals(obj:Any) = obj match {
    case f: FunNode => params.sameElements(f.params)
    case _ => false
  }
}

File: src/main/java/com/graphbrain/eco/nodes/FunNode.java
package com.graphbrain.eco.nodes;

public abstract class FunNode extends ListNode {

    public FunNode(ProgNode[] params, int lastTokenPos) {
        super(params, lastTokenPos);
    }

    public FunNode(ProgNode[] params) {
        this(params, -1);
    }

    public abstract String label();
}